### PR TITLE
DT-116 Add endpoint to create new case note types for secure/sensitiv…

### PR DESCRIPTION
**Pull request by Mjwillis**
…e case notes

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java [hunk 1]
```diff
@@ -114,4 +114,59 @@ public CaseNote amendCaseNote(
...
+            @ApiResponse(code = 409, message = "The case note sub type has already been recorded. The current unmodified object (including status) is returned.", response = ErrorResponse.class)})
+    public CaseNoteType createCaseNoteSubType(
+            @ApiParam(value = "Parent Case Note Type", required = true, example = "GEN") @PathVariable("parentType") final String parentType,
+            @RequestBody @NotNull final NewCaseNoteType body) {
+        return caseNoteService.createCaseNoteSubType(parentType, body);
+    }
+
+    @PutMapping(value = "/types/{parentType}", consumes = "application/json")
+    @ResponseStatus(HttpStatus.CREATED)
+    @ApiOperation(value = "Add New Case Note Type",
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java]: `Update New Case Not Type` also Response status = 200 ?

File: src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java [hunk 2]
```diff
@@ -114,4 +114,59 @@ public CaseNote amendCaseNote(
...
+            @ApiResponse(code = 201, message = "The Case Note Type has been recorded. The updated object is returned including the status.", response = CaseNoteType.class),
+            @ApiResponse(code = 404, message = "The case note type is not found", response = ErrorResponse.class)})
+    public CaseNoteType updateCaseNoteType(
+            @ApiParam(value = "Parent Case Note Type", required = true, example = "OBS") @PathVariable("parentType") final String parentType,
+            @RequestBody @NotNull final UpdateCaseNoteType body) {
+        return caseNoteService.updateCaseNoteType(parentType, body);
+    }
+
+    @PutMapping(value = "/types/{parentType}/{subType}", consumes = "application/json")
+    @ResponseStatus(HttpStatus.CREATED)
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java]: `OK` and value and notes can refer to update?

File: src/main/java/uk/gov/justice/hmpps/casenotes/dto/UpdateCaseNoteType.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@ApiModel(description = "Update a Case Note Type")
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@Data
+public class UpdateCaseNoteType {
+
+    @ApiModelProperty(required = true, value = "Type Description", example = "General Note Type", position = 2)
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/casenotes/dto/UpdateCaseNoteType.java]: position 1?

File: src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java [hunk 4]
```diff
@@ -285,4 +289,61 @@ public CaseNote getCaseNote(final String offenderIdentifier, final String caseNo
...
         return NumberUtils.isDigits(caseNoteIdentifier);
     }
+
+    @Transactional
+    @PreAuthorize("hasAnyRole('MAINTAIN_REF_DATA', 'SYSTEM_USER')")
+    public CaseNoteType createCaseNoteType(@Valid final NewCaseNoteType newCaseNoteType) {
+
+        final var parentNoteTypeOptional = parentCaseNoteTypeRepository.findById(newCaseNoteType.getType());
+
+        if (parentNoteTypeOptional.isPresent()) {
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java]: meh.. could get rid of temp variable.. 

```java
parentCaseNoteTypeRepository.findById(newCaseNoteType.getType()).orElseThrow(() -> new EntityExistsException(newCaseNoteType.getType())```

File: src/test/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteResourceTest.java [hunk 5]
```diff
@@ -321,4 +320,65 @@ public void testCanFilterCaseNotes() {
...
         assertJsonAndStatus(response, CaseNote.class, 200, "A1234AC-casenote.json");
     }
+
+    @Test
+    public void testCanCreateAndUpdateTypes() {
+        oauthMockServer.subGetUserDetails(SYSTEM_USER_READ_WRITE);
+
+        final var token = authTokenHelper.getToken(SYSTEM_USER_READ_WRITE);
+
+        // amend the case note
```
Comment (andrewrlee) [inline src/test/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteResourceTest.java]: are these comments correct?

**Comment by Mjwillis** [comment]
@andrewrlee updated based on your feedback

File: src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java (hunk 1 above)
Comment (andrewrlee) [reply src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java]: Think the value and notes still need to be updated?:
``` @ApiOperation(value = "Add New Case Note Sub Type",
            response = UpdateCaseNoteType.class,
            notes = "Creates a new case note sub type")
```
and below too

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/casenotes/controllers/CaseNoteController.java]: Duh!

**Pull request merged**